flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/AvroRowDeserializationSchema.java

```diff
@@ -0,0 +1,157 @@
+ * limitations under the License.
+ */
+package org.apache.flink.streaming.util.serialization;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.List;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.reflect.ReflectDatumReader;
+import org.apache.avro.specific.SpecificData;
+import org.apache.avro.specific.SpecificRecord;
+import org.apache.avro.util.Utf8;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * Deserialization schema from Avro bytes over {@link SpecificRecord} to {@link Row}.
+ *
+ * Deserializes the <code>byte[]</code> messages into (nested) Flink Rows.
+ *
+ * {@link Utf8} is converted to regular Java Strings.
+ */
+public class AvroRowDeserializationSchema extends AbstractDeserializationSchema<Row> {
+
+	/**
+	 * Schema for deterministic field order.
+	 */
+	private final Schema schema;
+
+	/**
+	 * Reader that deserializes byte array into a record.
+	 */
+	private final DatumReader<GenericRecord> datumReader;
+
+	/**
+	 * Input stream to read message from.
+	 */
+	private final MutableByteArrayInputStream inputStream;
+
+	/**
+	 * Avro decoder that decodes binary data
+	 */
+	private final Decoder decoder;
+
+	/**
+	 * Record to deserialize byte array to.
+	 */
+	private GenericRecord record;
+
+	/**
+	 * Creates a Avro deserialization schema for the given record.
+	 *
+	 * @param recordClazz Avro record class used to deserialize Avro's record to Flink's row
+	 */
+	@SuppressWarnings("unchecked")
+	public AvroRowDeserializationSchema(Class<? extends SpecificRecord> recordClazz) {
+		Preconditions.checkNotNull(recordClazz, "Avro record class must not be null.");
+		this.schema = SpecificData.get().getSchema(recordClazz);
+		this.datumReader = new ReflectDatumReader<>(schema);
+		this.record = new GenericData.Record(schema);
+		this.inputStream = new MutableByteArrayInputStream();
+		this.decoder = DecoderFactory.get().binaryDecoder(inputStream, null);
+	}
+
+	@Override
+	public Row deserialize(byte[] message) throws IOException {
+		// read record
+		try {
+			inputStream.setBuffer(message);
+			this.record = datumReader.read(record, decoder);
+		} catch (IOException e) {
+			throw new RuntimeException("Failed to deserialize Row.", e);
+		}
+
+		// convert to row
+		final Object row = convertToRow(schema, record);
+		return (Row) row;
+	}
+
+	/**
+	 * Converts a (nested) Avro {@link SpecificRecord} into Flink's Row type.
+	 * Avro's {@link Utf8} fields are converted into regular Java strings.
+	 */
+	private static Object convertToRow(Schema schema, Object recordObj) {
+		if (recordObj instanceof GenericRecord) {
+			// records can be wrapped in a union
+			if (schema.getType() == Schema.Type.UNION) {
```

Comment:
Not sure if we should support `UNION` at all. 
If the you have a UNION[NULL, RECORD] field in Avro, you'd expect it to be represented also as UNION field in a Table. 
We change it here to a nullable Record field. Not sure if that's expected.

Should we just not accept it (its a corner case anyway) and add support once the Table API / SQL support union types?